File: Cloude/Cloude/Services/ConnectionManager.swift
import Foundation
import Combine
import UIKit
import CloudeShared

@MainActor
class ConnectionManager: ObservableObject {
    @Published var connections: [UUID: EnvironmentConnection] = [:]

    let events = PassthroughSubject<ConnectionEvent, Never>()
    var conversationOutputs: [UUID: ConversationOutput] = [:]
    var conversationEnvironments: [UUID: UUID] = [:]
    private var backgroundTaskId: UIBackgroundTaskIdentifier = .invalid

    var isConnected: Bool { connections.values.contains { $0.isConnected } }
    var isAuthenticated: Bool { connections.values.contains { $0.isAuthenticated } }

    var agentState: AgentState {
        let states = connections.values.map(\.agentState)
        if states.contains(.running) { return .running }
        if states.contains(.compacting) { return .compacting }
        return .idle
    }

    var isAnyRunning: Bool {
        conversationOutputs.values.contains { $0.isRunning }
    }

    var lastError: String? {
        connections.values.compactMap(\.lastError).first
    }

    var processes: [AgentProcessInfo] {
        connections.values.flatMap(\.processes)
    }

    var skills: [Skill] {
        var seen = Set<String>()
        return connections.values.flatMap(\.skills).filter { seen.insert($0.name).inserted }
    }

    var isWhisperReady: Bool {
        connections.values.contains { $0.isWhisperReady }
    }

    var isTranscribing: Bool {
        connections.values.contains { $0.isTranscribing }
    }

    var defaultWorkingDirectory: String? {
        connections.values.compactMap(\.defaultWorkingDirectory).first
    }

    var chunkProgress: EnvironmentConnection.ChunkProgress? {
        connections.values.compactMap(\.chunkProgress).first
    }

    var fileCache: AggregateFileCache {
        AggregateFileCache(connections: connections)
    }

    func output(for conversationId: UUID) -> ConversationOutput {
        if let existing = conversationOutputs[conversationId] {
            return existing
        }
        let new = ConversationOutput()
        new.parent = self
        conversationOutputs[conversationId] = new
        return new
    }

    func connection(for environmentId: UUID?) -> EnvironmentConnection? {
        environmentId.flatMap { connections[$0] }
    }

    func connectionForConversation(_ conversationId: UUID) -> EnvironmentConnection? {
        conversationEnvironments[conversationId].flatMap { connections[$0] }
    }

    func anyAuthenticatedConnection() -> EnvironmentConnection? {
        connections.values.first { $0.isAuthenticated }
    }

    func connectEnvironment(_ envId: UUID, host: String, port: UInt16, token: String, symbol: String = "laptopcomputer") {
        let conn = connections[envId] ?? EnvironmentConnection(environmentId: envId)
        conn.manager = self
        conn.symbol = symbol
        connections[envId] = conn
        conn.connect(host: host, port: port, token: token)
    }

    func disconnectEnvironment(_ envId: UUID, clearCredentials: Bool = true) {
        connections[envId]?.disconnect(clearCredentials: clearCredentials)
    }

    func disconnectAll(clearCredentials: Bool = true) {
        for conn in connections.values {
            conn.disconnect(clearCredentials: clearCredentials)
        }
    }

    func reconnectAll() {
        for conn in connections.values {
            conn.reconnectIfNeeded()
        }
    }

    func registerConversation(_ conversationId: UUID, environmentId: UUID) {
        conversationEnvironments[conversationId] = environmentId
    }

    func clearAllRunningStates() {
        for output in conversationOutputs.values {
            for i in output.toolCalls.indices where output.toolCalls[i].state == .executing {
                output.toolCalls[i].state = .complete
            }
            output.isRunning = false
            output.isCompacting = false
        }
        for conn in connections.values {
            conn.agentState = .idle
            conn.runningConversationId = nil
        }
    }

    func beginBackgroundStreamingIfNeeded() {
        guard isAnyRunning, backgroundTaskId == .invalid else { return }
        backgroundTaskId = UIApplication.shared.beginBackgroundTask(withName: "StreamingResponse") { [weak self] in
            self?.endBackgroundStreaming()
        }
    }

    func endBackgroundStreaming() {
        guard backgroundTaskId != .invalid else { return }
        let taskId = backgroundTaskId
        backgroundTaskId = .invalid
        UIApplication.shared.endBackgroundTask(taskId)
    }

    func fileCache(for environmentId: UUID?) -> FileCache? {
        environmentId.flatMap { connections[$0]?.fileCache }
    }
}

File: android/app/src/main/java/com/cloude/app/Services/ConnectionManager.kt
package com.cloude.app.Services

import com.cloude.app.Models.AgentState
import com.cloude.app.Models.ClientMessage
import com.cloude.app.Models.ServerEnvironment
import com.cloude.app.Models.ServerMessage
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow

class ConnectionManager(private val appContext: android.content.Context? = null) {
    private val connections = mutableMapOf<String, EnvironmentConnection>()
    private val conversationOutputs = mutableMapOf<String, ConversationOutput>()
    private val conversationEnvironments = mutableMapOf<String, String>()

    private val _isConnected = MutableStateFlow(false)
    val isConnected: StateFlow<Boolean> = _isConnected

    private val _isAuthenticated = MutableStateFlow(false)
    val isAuthenticated: StateFlow<Boolean> = _isAuthenticated

    private val _agentState = MutableStateFlow(AgentState.idle)
    val agentState: StateFlow<AgentState> = _agentState

    private val _events = MutableSharedFlow<ServerMessage>(extraBufferCapacity = 64)
    val events: SharedFlow<ServerMessage> = _events

    val isAnyRunning: Boolean
        get() = conversationOutputs.values.any { it.isRunning.value }

    fun output(conversationId: String): ConversationOutput =
        conversationOutputs.getOrPut(conversationId) { ConversationOutput() }

    fun registerConversation(conversationId: String, environmentId: String) {
        conversationEnvironments[conversationId] = environmentId
    }

    fun connectionForConversation(conversationId: String): EnvironmentConnection? =
        conversationEnvironments[conversationId]?.let { connections[it] }

    fun connection(forId: String): EnvironmentConnection? = connections[forId]

    fun connectEnvironment(env: ServerEnvironment) {
        val existing = connections[env.id]
        if (existing != null) {
            existing.disconnect()
        }

        val conn = EnvironmentConnection(env.id, appContext) { message ->
            _events.tryEmit(message)
            updateAggregateState()
        }
        connections[env.id] = conn
        conn.connect(env.host, env.port, env.token)
        updateAggregateState()
    }

    fun disconnectEnvironment(envId: String) {
        connections.remove(envId)?.disconnect()
        updateAggregateState()
    }

    fun disconnectAll() {
        connections.values.forEach { it.disconnect() }
        connections.clear()
        updateAggregateState()
    }

    fun reconnectAll() {
        connections.values.forEach { it.reconnect() }
    }

    fun send(message: ClientMessage, environmentId: String) {
        val conn = connections[environmentId]
        if (conn == null) {
            android.util.Log.e("Cloude", "No connection for envId=$environmentId, available=${connections.keys}")
            return
        }
        conn.send(message)
    }

    private fun updateAggregateState() {
        _isConnected.value = connections.values.any { it.isConnected.value }
        _isAuthenticated.value = connections.values.any { it.isAuthenticated.value }
        _agentState.value = connections.values
            .map { it.agentState.value }
            .firstOrNull { it != AgentState.idle }
            ?: AgentState.idle
    }
}
